File: oauth2/src/jvmTest/java/com/okta/oauth2/kmp/jvm/ResourceOwnerFlowTest.java
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp.jvm;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import com.okta.authfoundation.client.TokenInfo;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import org.junit.Test;

/**
 * Java-language tests for {@link ResourceOwnerFlow} CompletableFuture wrapper. Validates that the
 * API is ergonomic from actual Java code (FR-016).
 */
public class ResourceOwnerFlowTest {

  @Test
  public void start_ReturnsCompletableFutureWithTokenInfo()
      throws ExecutionException, InterruptedException, TimeoutException {
    ResourceOwnerFlow flow = TestFlowFactory.createSuccessResourceOwnerFlow();
    try {
      CompletableFuture<TokenInfo> future = flow.start("user@example.com", "password", "openid");
      assertNotNull("Future should not be null", future);

      TokenInfo tokenInfo = future.get(5, TimeUnit.SECONDS);
      assertNotNull("TokenInfo should not be null", tokenInfo);
      assertEquals("Bearer", tokenInfo.getTokenType());
      assertEquals("test-access-token", tokenInfo.getAccessToken());
    } finally {
      flow.close();
    }
  }

  @Test
  public void start_WithError_CompletesExceptionally()
      throws TimeoutException, InterruptedException {
    ResourceOwnerFlow flow = TestFlowFactory.createFailingResourceOwnerFlow();
    try {
      CompletableFuture<TokenInfo> future = flow.start("user@example.com", "wrong", "openid");
      try {
        future.get(5, TimeUnit.SECONDS);
        fail("Should have thrown ExecutionException");
      } catch (ExecutionException e) {
        assertNotNull("Cause should not be null", e.getCause());
        assertTrue(
            "Should contain error message", e.getCause().getMessage().contains("invalid_grant"));
      }
    } finally {
      flow.close();
    }
  }

  @Test
  public void close_IsIdempotent() {
    ResourceOwnerFlow flow = TestFlowFactory.createSuccessResourceOwnerFlow();
    flow.close();
    flow.close(); // Should not throw
  }
}

File: oauth2/src/commonTest/kotlin/com/okta/oauth2/kmp/ResourceOwnerFlowTest.kt
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp

import com.okta.authfoundation.client.OAuth2ClientBuilder
import com.okta.authfoundation.client.TokenInfo
import com.okta.authfoundation.client.kmp.OAuth2Client
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class ResourceOwnerFlowTest {
    private val fakeClient: OAuth2Client =
        OAuth2ClientBuilder
            .create(
                issuerUrl = "https://example.okta.com",
                clientId = "test-client-id",
                scope = listOf("openid")
            ).getOrThrow()

    @Test
    fun start_WithValidCredentials_ReturnsTokenInfo() =
        runTest {
            val capturedParams = mutableMapOf<String, String?>()
            val mockFlow =
                object : ResourceOwnerFlow {
                    override val client: OAuth2Client = fakeClient

                    override suspend fun start(
                        username: String,
                        password: String,
                        scope: List<String>,
                    ): Result<TokenInfo> {
                        capturedParams["username"] = username
                        capturedParams["password"] = password
                        capturedParams["scope"] = scope.joinToString(" ")
                        return Result.success(FakeTokenInfo())
                    }
                }

            val result = mockFlow.start("user@example.com", "secret", listOf("openid", "profile"))

            assertTrue(result.isSuccess)
            assertNotNull(result.getOrNull())
            assertEquals("user@example.com", capturedParams["username"])
            assertEquals("secret", capturedParams["password"])
            assertEquals("openid profile", capturedParams["scope"])
        }

    @Test
    fun start_WithInvalidCredentials_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : ResourceOwnerFlow {
                    override val client: OAuth2Client = fakeClient

                    override suspend fun start(
                        username: String,
                        password: String,
                        scope: List<String>,
                    ): Result<TokenInfo> = Result.failure(IllegalArgumentException("invalid_grant"))
                }

            val result = mockFlow.start("user@example.com", "wrong", listOf("openid"))

            assertTrue(result.isFailure)
            assertEquals("invalid_grant", result.exceptionOrNull()?.message)
        }

    @Test
    fun start_WithNetworkError_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : ResourceOwnerFlow {
                    override val client: OAuth2Client = fakeClient

                    override suspend fun start(
                        username: String,
                        password: String,
                        scope: List<String>,
                    ): Result<TokenInfo> = Result.failure(java.io.IOException("Network unreachable"))
                }

            val result = mockFlow.start("user@example.com", "pass", listOf("openid"))

            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull() is java.io.IOException)
        }
}
